apollo-router/src/http_subgraph.rs

```diff
@@ -84,6 +92,60 @@ impl HttpSubgraphFetcher {
             })
         })
     }
+
+    async fn dedup(
+        &self,
+        request: graphql::Request,
+    ) -> Result<graphql::Response, graphql::FetchError> {
+        loop {
+            let mut locked_wait_map = self.wait_map.lock().await;
+            match locked_wait_map.get_mut(&request) {
+                Some(waiter) => {
+                    // Register interest in key
+                    let mut receiver = waiter.subscribe();
+                    drop(locked_wait_map);
+
+                    match receiver.recv().await {
+                        Ok(value) => return value,
+                        // there was an issue with the broadcast channel, retry fetching
+                        Err(_) => continue,
+                    }
+                }
+                None => {
+                    let (tx, _rx) = broadcast::channel(1);
+                    locked_wait_map.insert(request.clone(), tx.clone());
+                    drop(locked_wait_map);
+
+                    let res = self.fetch(request.clone()).await;
+
+                    {
+                        let mut locked_wait_map = self.wait_map.lock().await;
+                        locked_wait_map.remove(&request);
+                    }
+
+                    // Let our waiters know
+                    let broadcast_value = res.clone();
+                    // Our use case is very specific, so we are sure that
+                    // we won't get any errors here.
+                    tokio::task::spawn_blocking(move || {
+                    tx.send(broadcast_value)
+                        .expect("there is always at least one receiver alive, the _rx guard; qed")
+                }).await
+                .expect("can only fail if the task is aborted or if the internal code panics, neither is possible here; qed");
+                    return res;
+                }
+            }
```

Comment:
I'm definitely beginning to feel there is some kind of generic functionality which we can extract for common use. I might try and prototype something up...

Reply:
yup, that could be generalized